File: GO/Heap sort.go

  
func heapify(items []int, idx int, size int) {  
	l := 2*idx + 1 // left = 2*i + 1  
	r := 2*idx + 2 // right = 2*i + 2  
   
	var largest int;  
	if l <= size && items[l] > items[idx] {  
	   largest = l  
	} else {  
	   largest = idx  
	}  
   
	if r <= size && items[r] > items[largest] {  
	   largest = r  
	}  
   
	if largest != idx {  
	   t := items[idx]  
	   items[idx] = items[largest]  
	   items[largest] = t  
	   heapify(items, largest, size)  
	}  
   
 }  
   
 func HeapSort(items []int) {  
   
	L := len(items) //heap size  
	m := int(L / 2) //middle  
   
	for i := m; i >= 0; i-- {  
   
	   heapify(items, i, L-1)  
	}  
   
	F := L - 1  
	for j := F; j >= 0; j-- {  
	   t := items[0]  
	   items[0] = items[j]  
	   items[j] = t  
	   F -= 1  
	   heapify(items, 0, F)  
	}  
   
 }  
 
 
 

File: Python/heapsort.py
def heap(arr, n, i):
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[i] < arr[left]:
        largest = left

    if right < n and arr[largest] < arr[right]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heap(arr, n, largest)

def heapSort(arr):
    n = len(arr)

    for i in range(n//2 - 1, -1, -1):
        heap(arr, n, i)

    for i in range(n-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heap(arr, i, 0)

arr = [11, 10, 14, 6, 7, 8]
heapSort(arr)
print("Sorted array is", arr)
